### ISSUE TITLE
回归实验的实验日志

### ISSUE THREAD

**Issue by WritingEve** [XuanhaoLiu/EEG2Video#12]
作者您好！非常感谢您所做的杰出的工作！我刚刚入门，在复现时，有几个问题：
1.train_semantic_predictor以及transformer预测vae潜变量时，能否提供一下具体的实验训练日志呢？我自己复现时，采用sub1的de1_per1s的数据和1~6的txt文本数据，用txt7做eval时，回归loss在6左右浮动。我很想确定训练过程是否正确。
2.针对每一个sub都需要训练一个train_semantic_predictor吗？不同sub均需要用于后续的video_reconstruction？好像论文里并没有报道不同sub的准确率
非常期待您的解答！！！

**Comment by XuanhaoLiu** [comment]
您好！感谢您对我们工作的关注！

1. transformer预测vae潜变量时，在训练集上收敛，在测试集上loss在6浮动是正常的；因为对于脑电信号直接对齐到视频的图像信息本身就是一件很难的事情，所以直接去对齐在测试集上loss大是合理的。
 
2. train_semantic_predictor对齐text的clip embedding需要收敛到更小的精度，否则无法指导扩散模型生成正确的视频。具体来说，你可以减小learning rate，加大训练次数，使mse loss收敛到0.1以下才行。你可以使用一些技巧训练模型，即使用人造数据集先得到预训练权重，再在此基础上使用脑电信号继续训练以对齐text的clip embedding。人造数据集即为更简单的数据集：比如原来的数据是{EEG_data, txt_emb}, 现在将所有同一类的数据均使用一个更简单fake_data = torch.ones_like(EEG_data) * class_label，这样得到一个假的人造数据集{fake_data, txt_emb}。先在人造数据集上训练使得train_semantic_predictor收敛一部分，再继续在SEED-DV上训练使得train_semantic_predictor收敛至训练集。

3. 是的，每一个sub都需要训练一个train_semantic_predictor，因为目前的brain decoding做不到真正意义上的跨被试。

4. 由于时间匆忙，论文里只汇报了sub 1的结果，并没有跑后续被试的重建效果。

希望这些回答可以帮助你解决你的问题！祝好！

**Comment by ZhiShengZhou1973** [comment]
两位，你们好！
我在复现EEG信号和文本信号对齐时，出现了一下的问题，想请教您们是否也遇到这样的问题，以及是否有解决方法。
1. 采用作者您提出的{fake_data,txt_emb}的训练方法后，我现在人造数据集训练，再切换到SEED-DV数据集后，test loss 反而会变大且很难下降，不像作者您说的可以继续收敛。所以想知道您说的收敛是收敛到训练集还是测试集。
2. 除了采用fake data 的训练方法，您们还有尝试过什么训练策略来解决，由于数据集数据太少导致无法对齐的问题吗

**Comment by WritingEve** [comment]
感谢作者的回答！我复现时还遇到了另外几个问题：
1.您的inference代码中是怎样体现dynamic_predictor的呢？根据您的论文，您模型是根据slow\fast来选择采样的beta取值的，我好像没有看到这部分代码——只有Seq2Seq\Text对齐的数据代码
2.我想询问一下您论文中关于tuneAvideo的训练细节。原本的tuneAvideo训练是一个视频训练，5条prompt验证；您的代码中数据集包含200条视频——我使用200条视频全部当作训练集发现效果非常差，生成出来的视频我推测模型完全无法当作inference的基座。所以我想询问一下您这部分模型训练时数据集该如何选取？是每一类训练还是每一条视频单独训练还是有另外的数据集划分。
3.另外您提到的fakedata，我实验发现在训练集上确实能够下降到0.01甚至更低，但是在测试集上还是难以下降。
期待作者您的回答！祝您科研顺利！！！

**Comment by XuanhaoLiu** [comment]
> 两位，你们好！ 我在复现EEG信号和文本信号对齐时，出现了一下的问题，想请教您们是否也遇到这样的问题，以及是否有解决方法。
> 
> 1. 采用作者您提出的{fake_data,txt_emb}的训练方法后，我现在人造数据集训练，再切换到SEED-DV数据集后，test loss 反而会变大且很难下降，不像作者您说的可以继续收敛。所以想知道您说的收敛是收敛到训练集还是测试集。
> 2. 除了采用fake data 的训练方法，您们还有尝试过什么训练策略来解决，由于数据集数据太少导致无法对齐的问题吗

1. 是训练集。测试集上收敛是一件很难的事情，fMRI dataset也做不到这一点。
2. 减少prompt的种类数，比如 a plane is flying 和 a plane is flying in the sky 可以统一成一个prompt。比如都用后者。不过总归来讲，这种方式治标不治本，根本问题还是数据集数据太少导致对齐困难。

**Comment by XuanhaoLiu** [comment]
> 感谢作者的回答！我复现时还遇到了另外几个问题： 1.您的inference代码中是怎样体现dynamic_predictor的呢？根据您的论文，您模型是根据slow\fast来选择采样的beta取值的，我好像没有看到这部分代码——只有Seq2Seq\Text对齐的数据代码 2.我想询问一下您论文中关于tuneAvideo的训练细节。原本的tuneAvideo训练是一个视频训练，5条prompt验证；您的代码中数据集包含200条视频——我使用200条视频全部当作训练集发现效果非常差，生成出来的视频我推测模型完全无法当作inference的基座。所以我想询问一下您这部分模型训练时数据集该如何选取？是每一类训练还是每一条视频单独训练还是有另外的数据集划分。 3.另外您提到的fakedata，我实验发现在训练集上确实能够下降到0.01甚至更低，但是在测试集上还是难以下降。 期待作者您的回答！祝您科研顺利！！！

我们会尽快公布新代码。
1. 加噪的时候，使用slow\fast 决定beta值，根据beta值加噪。
2. 具体的情况我也不太记得了。您可以先用一些10类的子集来训练＋推理。
3. 实在训练集上收敛，测试集收敛是很困难的，fMRI dataset也很难做到。

**Comment by aWeirdName** [comment]
作者你好，可以解释一下语义对齐中     for i in range(6):
        # Text.append(text_embedding[:30,...])
        Text.append(text_embedding[:150,...])
    Text = np.concatenate(Text)   文本表征这样处理的原因么？ 这个text_embedding原本应该是什么维度，对应怎么样排列的eeg信号。为什么是前150个重复六次呢。

**Comment by aWeirdName** [comment]
此外这个数据集自带的label也就是脚本里的label究竟是什么为什么这么排列。

**Comment by aWeirdName** [comment]
此外这个数据集自带的label也就是脚本里的label究竟是什么为什么这么排列。您数据集中sub1看起来是重复了两次实验，但是sub1和sub1_session2对应的数据我发现差距非常的大，可以解释一下sub1_session2是什么数据吗。

![Image](https://github.com/user-attachments/assets/b8b6cf0f-a4b7-4bf5-a7d9-7fb479055abe)

![Image](https://github.com/user-attachments/assets/c5cc0008-6371-45d6-85d6-990dcf7c2c1a)

**Comment by XuanhaoLiu** [comment]
我们已经公布新代码！

**Comment by WritingEve** [comment]
感谢作者您的更新代码！我在阅读新代码的过程中，有以下疑惑：
1.在tuneAvideo的代码中，您的config里关于train-data参数仍然是单个视频"block1/163.mp4"，但是您训练代码中使用的仍然是TuneMultiVideo，我的理解是您训练时使用的是多个视频当作训练集（而非tuneAvideo源代码中的单个视频），请问这里该如何处理呢？
2.关于Seq2Seq任务里，latent_data(1200就是6*200段视频的vae变量吗?)，另外test-latent的40_classes.npy是如何得来的呢？(是否就是40类别的prompt模板？)
期待作者您的解答！！！祝您科研顺利！

**Comment by KevinZhou6** [comment]
感谢您对我们项目的支持！
问题1：可以参考`train_finetune_videodiffusion.py`中第185-209行以及*TuneMultiVideoDataset*的代码，在真实训练过程中，通过手动替换的方式将训练集变成多个视频，而非原本Tune-A-Video工作的单个视频。
问题2：无论是`1200_latent.npy`还是`40classes_latents.pt`都是VAE变量。我们选择block7的数据作为测试集，block1-6作为训练集。
祝好！

**Comment by WritingEve** [comment]
作者您好！感谢您前面的回答，我将实验大致流程跑通了，但却无法得到正确的生成结果，有几个疑问想问您：
1.训练videodiff时，模型的训练集是6个block共1200段视频吗？将这些全部当作训练集进行训练？
2.用于inference时的eeg应该是由eeg_text的clip模型得到的，但是我使用该eeg却无法生成正确的视频，请问这一部分的处理细节是怎样的呢？
3.用于seq2seq的模型，我的loss只能降到18左右，请问为了降到6左右还有哪些处理细节呢？
期待作者的回答！祝您科研顺利！

**Comment by XuanhaoLiu** [comment]
感谢您对我们工作的关注：
1. 训练video diff的时候，不需要用那么多视频，每个种类选1-2个视频即可，您现在可以前往MindCine论文的github和huggingface下载预训练的video diff模型权重。
2. 你需要将eeg的text encoder的训练loss降至0.1以下，需要使用更少一些的数据规格，比如每一个种类的视频标题原来最多有35个，现在只保留1个。以及使用伪数据集进行预训练，具体伪数据集的预训练代码在MindCros的training.py可以看到。
3. seq2seq的模型，确实testing loss会很大，但只要在training set上收敛即可。
